src/toxcore/tcp/server/server.rs

```diff
@@ -105,88 +106,120 @@ impl Server {
+
+            // check if client_a is already linked
+            if let Some(index) = links_a.id_by_pk(&packet.pk) {
+                // send RouteResponse if client was already linked to pk
+                return client_a.send_route_response(&packet.pk, index + 16)
+            }
+
+            // try to insert a new link
+            if let Some(index) = links_a.insert(&packet.pk) {
+                index
             } else {
-                return Box::new(future::err(
-                    Error::new(ErrorKind::Other,
-                               "RouteRequest: no such PK"
-                    )))
+                // send RouteResponse(0) if no space to insert new link
+                return client_a.send_route_response(&packet.pk, 0)
             }
         };
-        let client_a = &state.connected_clients[pk];
-        if let Some(client_b) = state.connected_clients.get(&packet.pk) {
-            // check if current pk is linked inside other_client
-            if let Some(a_id_in_client_b) = client_b.get_connection_id(pk) {
-                // the are both linked, send RouteResponse and
-                // send each other ConnectNotification
-                // we don't care if connect notifications fail
-                let client_a_notification = client_a.send_connect_notification(b_id_in_client_a);
-                let client_b_notification = client_b.send_connect_notification(a_id_in_client_b);
-                return Box::new(
-                    client_a.send_route_response(&packet.pk, b_id_in_client_a)
-                        .join(client_a_notification)
-                        .join(client_b_notification)
-                        .map(|_| ())
-                )
+
+        // get client_b
+        let client_b = if let Some(client) = state.connected_clients.get(&packet.pk) {
+            client
+        } else {
+            // send RouteResponse only to current client
+            return client_a.send_route_response(&packet.pk, b_id_in_client_a + 16)
+        };
+
+        // check if current pk is linked inside other_client
+        let a_id_in_client_b = {
+            if let Some(index) = client_b.links().id_by_pk(&packet.pk) {
+                index
             } else {
                 // they are not linked
                 // send RouteResponse only to current client
-                client_a.send_route_response(&packet.pk, b_id_in_client_a)
+                return client_a.send_route_response(&packet.pk, b_id_in_client_a + 16)
             }
-        } else {
-            // send RouteResponse only to current client
-            client_a.send_route_response(&packet.pk, b_id_in_client_a)
-        }
+        };
+
+        // the are both linked, send RouteResponse and
+        // send each other ConnectNotification
+        // we don't care if connect notifications fail
+        client_a.links().upgrade(b_id_in_client_a);
+        client_b.links().upgrade(a_id_in_client_b);
```

Comment:
It seems it might cause a deadlock if two clients send RouteRequest simultaneously and processed in parallel.

Reply:
But there might be another problem. What if one client sends ConnectNotification whereas another client sends DisconnectNotification? In this case links might become inconsistent - client_a will have online link but client_b will have no link.

Reply:
> In this case links might become inconsistent - client_a will have online link but client_b will have no link.

Yes there is no solution for that but to make a single threaded tcp server